core/includes/form.inc

```diff
@@ -4986,6 +5004,96 @@ function _form_set_class(&$element, $class = array()) {
   }
 }
 
+/**
+ * Checks the number of input form elements to ensure that the PHP
+ * max_input_vars limit is not exceeded.
+ *
+ * @param array $count
+ *   The number of elements in the form.
+ * @param string $error_level
+ *   The error level for the message displayed on screen if the input vars value
+ *   has been exceeded. Either 'warning' or 'error'.
+ *
+ * @see form_pre_render()
+ */
+function _form_check_input_vars($count, $error_level = 'warning') {
+  // Determine the limit on input vars for this server configuration.
+  $limit = ini_get('max_input_vars');
+  if ($limit) {
+    // Leave a 5% buffer for ajax elements or similar.
+    if ($count > $limit * 0.95) {
+      $substitutions = array(
+        '@count' => $count,
+        '@limit' => $limit,
+      );
+
+      // Translatable strings.
+      $warning = t('This form may be too long to work properly.');
+      $notice = t('The last form processed was longer than this system can safely support.');
+      $please_increase = t('This form contains @count input elements. PHP <code>max_input_vars</code> allows only @limit and should be increased.', $substitutions);
+      $data_loss = t('The form has been submitted, but data may heve been lost.');
+
+      switch ($error_level) {
+        case 'error':
+          if (user_access('administer site configuration')) {
+            $message = $data_loss . ' ' . $please_increase;
+          }
+          else {
+            $message = $notice . ' ' . $data_loss;
+          }
+          break;
+
+        // 'warning'
+        default:
+          if (user_access('administer site configuration')) {
+            $message = $please_increase;
+          }
+          else {
+            $message = $warning;
+          }
+      }
+
+      // Remove existing messages that could be confusing.
+      $set_new_message = TRUE;
+      if (!empty($_SESSION['messages'])) {
+        $current_path = current_path();
+        $translations = array(
+          'admin/config/people/permissions' => t('The changes have been saved.'),
+          'admin/structure/taxonomy' => t('The configuration options have been saved.'),
+          'admin/structure/menu' => t('Your configuration has been saved.'),
+        );
```

Comment:
Is there any better way to do this? (instead of duplicating anticipated text here) 🤔 ...if these messages happen to change, then we need to somehow remember to change them here too. ...is there a way to reference them?

Thoughts/ideas:
- use constants (seems like a messy workaround and I don't like it)
- a helper function that generates these messages based on path (passed as a param to the function), and then we reuse that function here, as well as in the respective submit handlers for the forms (those that are prone to this issue). The function would be generating two sets of messages: one for when `max_input_vars` is sufficient, and one for when not.
- something that intercepts these messages in `backdrop_set_message()`? 🤔